tests/phpunit/Unit/Graph/GraphFormatterTest.php

```diff
@@ -16,89 +16,225 @@
+					[ 'predicate' => 'Part of Team', 'object' => 'Team:Alpha' ],
+				], 'fields' => [
+					[ 'name' => 'Rated as', 'value' => 20, 'type' => '_num', 'page' => 'Rating' ]
+				] ]
+			],
+			'legend' => '<div class="graphlegend">' .
+				'<div class="graphlegenditem" style="color: black">black: Casted</div>' .
+				'<div class="graphlegenditem" style="color: red">red: Part of Team</div>' .
+				'</div>',
+			'dot' => <<<'FIELDS'
+digraph "Unit Test" {graph [fontsize=10, fontname="Verdana"]
+node [fontsize=10, fontname="Verdana"];
+edge [fontsize=10, fontname="Verdana"];
+size="100";node [shape=rect];rankdir=LR;
+"Team:Alpha" [label = <
+<table border="0" cellborder="0" cellspacing="1" columns="*" rows="*">
+<tr><td colspan="2" href="[[Team:Alpha]]">Alpha</td></tr><hr/>
+<tr><td align="left" href="[[Property:Rating]]">Rated as</td><td align="right">10</td></tr>
+</table>
+>, tooltip = "Alpha"];
+"Team:Beta" [label = <
+<table border="0" cellborder="0" cellspacing="1" columns="*" rows="*">
+<tr><td colspan="2" href="[[Team:Beta]]">Beta</td></tr><hr/>
+<tr><td align="left" href="[[Property:Rating]]">Rated as</td><td align="right">20</td></tr>
+</table>
+>, tooltip = "Beta"];
+"Person:Alexander Gesinn" -> "Team:Alpha" [label="Casted",fontcolor=black,arrowhead=diamond,color=black];
+"Person:Sebastian Schmid" -> "Team:Beta" [label="Casted",fontcolor=black,arrowhead=diamond,color=black];
+"Person:Alexander Gesinn" -> "Team:Beta" [label="Casted",fontcolor=black,arrowhead=diamond,color=black];
+"Team:Alpha" -> "Team:Beta" [label="Part of Team",fontcolor=red,arrowhead=diamond,color=red];
+}
+FIELDS
+		]
+	];
+
+	/** @const array BASE_PARAMS A non-changing subset of parameters. */
+	/** @see https://www.semantic-mediawiki.org/wiki/Help:Graph_format */
+	private const BASE_PARAMS = [
+		'graphname' => 'Unit Test',
+		'graphsize' => '100',
+		'graphfontsize' => 10,
+		'nodeshape' => 'rect',
+		'nodelabel' => 'displaytitle',
+		'arrowdirection' => 'LR',
+		'arrowhead' => 'diamond',
+		'wordwraplimit' => 20,
+		'relation' => 'parent',
+		'graphlink' => true,
+		'graphlabel' => true,
+		'graphcolor' => true,
+		'graphlegend' => true,
+	];
+
+	/**
+	 * Create a complete graph for the test case.
+	 * @var array $case
+	 * @return GraphFormatter
+	 */
+	private static function graph( array $case ): GraphFormatter {
+		$graph = new GraphFormatter( new GraphOptions( GraphFormatterTest::BASE_PARAMS + $case['params'] ) );
+		$nodes = [];
+		foreach ( $case['nodes'] as $node ) {
+			$graph_node = new GraphNode( $node['name'] );
+			$graph_node->setLabel( $node['label'] );
+			if ( isset( $node['parents'] ) ) {
+				foreach ( $node['parents'] as $parent ) {
+					$graph_node->addParentNode( $parent['predicate'], $parent['object'] );
+				}
+			}
+			if ( isset( $node['fields'] ) ) {
+				foreach ( $node['fields'] as $field ) {
+					$graph_node->addField( $field['name'], $field['value'], $field['type'], $field['page'] );
```

Comment:
Control structures in tests are a sign of something going wrong. Been a while since I saw a case with trippy nested control structures.